[smsmodem] Initial contribution

dalgwen opened the pull request:
This binding connects to a USB serial GSM modem (or a network exposed one, a.k.a ser2net) and allows openHAB to send and receive SMS through it.
With this binding, there is no need for complex configuration with external tools, such as gammu.

[Here is the community discussion](https://community.openhab.org/t/smsmodem-binding-for-sending-and-receiving-sms-on-a-gsm-serial-modem/128412), where you can find a binary snapshot.

This binding uses smslibv4, directly included (org.openhab.binding.smsmodem.internal.smslib packages and sub packages) and heavily refactored to pass openHAB code standard. I [asked on the forum if it was OK](https://community.openhab.org/t/inclusion-of-a-very-slightly-modified-library-in-the-source-code-of-a-binding-and-subsequent-issue/127597/2) and decided to go on with the inclusion. Hope this is OK. Honestly, I don't understand everything in this library : SMS is a complex standard and I didn't take the time to grab it as the smslibv4 was here for me to hide the complexity.

The code still has a "deviation" from the openHAB standard : in the internal smslib, it uses a manually created Thread to continuously read from the serial port (the PollReader class). As the thread is continuously running, I thought it could be a valid use case (to not take forever a thread from the openHAB pool). If you don't think so, please tell me and I will refactor it.

Hilbrand commented:
There are other add-ons that include third party code. But because it's under a different license this needs to be added differently. You need to put the sources in a different map: `src/3thparty/java/` and than you can keep the original package and such, and add the directory to the `pom.xml`. See for example the smartmeter binding: https://github.com/openhab/openhab-addons/tree/main/bundles/org.openhab.binding.smartmeter. Don't

The way you have added it now seems to me illegal as you have changed the license on the 3rd party source code. Which you only can do if you are the copyright owner of the source code.

Regarding threads. It's ok to have an own thread. In fact if you have long running threads you must create them as only short running threads are allowed on the scheduled executor. See also #8216 about naming the thread and do `setDeamon(true)`.

dalgwen commented:
Hello @Hilbrand,

Thank you very much for these informations.
Indeed, the way I included the code from smslib was problematic and I'm glad you directed me to a valid solution.
So, I did as you said, and refactored the inclusion. For information :
- I put back the smslib code in the org.smslib package
- I put this package and subpackages in the src/3rdparty/java folder
- I deleted the openHAB copyright information I had improperly added on top of every org.smslib file (though I let the mention "extracted from smslib" that I had put before).
- I completed the NOTICE file and the pom.xml with relevant information (I took the smartmeter binding as a model)

I also put a name and the daemon property on the always living thread polling the device.

wborn commented on bundles/org.openhab.binding.smsmodem/pom.xml:
```diff
@@ -0,0 +1,39 @@
+  <parent>
+    <groupId>org.openhab.addons.bundles</groupId>
+    <artifactId>org.openhab.addons.reactor.bundles</artifactId>
+    <version>3.3.0-SNAPSHOT</version>
```
```suggestion
    <version>3.4.0-SNAPSHOT</version>
```

dalgwen replied:
Done, by rebasing

fwolter commented:
@dalgwen It's been a long time you submitted your new addon. Are you still around to implement upcoming review feedback?

dalgwen commented:
Yes, I would be happy to do so.
I use this binding every day and also provide support on the forum for the (very) few interested users.

But I'm aware that it doesn't attract many people and is low on the priority list... so no problem if it takes time !

Thanks,

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
```diff
@@ -0,0 +1,129 @@
+
+This binding connects to a USB serial GSM modem (or a network exposed one, see ser2net) and allows openHAB to send and receive SMS through it.
+
+Serial modem should all use the same communication protocol (AT message) and therefore this binding _should_ be compatible with every dongle. However, there is a gap between theory and reality and success may vary. The protocol stack is based on the no longer supported smslib project (more precisely a v4 fork), and all modems supported by this library should be OK. The following devices have been ...
```
If you put each sentence in a separate line, reading the diffs would be easier.

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
```diff
@@ -0,0 +1,129 @@
+There is no discovery process for *smsmodembridge* thing.
+A *smsconversation* thing will be discovered and added to the inbox everytime the modem should receive a SMS by a new sender.
+
+## Thing Configuration
+
+The *smsmodembridge* thing requires at least two parameters to work properly (serialPortOrIP, baudOrNetworkPort).
+Depending on the nature of the connection (direct serial modem, or serial over network), this two fields will be used differently :
+
+| field | direct serial modem   | serial over network                  |
+|-------|--------------------------|-----------------------------------|
+|serialPortOrIP| The serial port to access (eg. /dev/tty/USBx) | IP address of the computer hosting the ser2net service|
+|baudOrNetworkPort| Baud rate                                 | The network port of the ser2net service |
```
Did you consider creating two bridge types? One for direct connection and another for ser2net?

fwolter replied:
I think it would be better from the usability perspective, but also from the modelling. But I won't insist on changing it.

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
```diff
@@ -0,0 +1,129 @@
+
+The other parameters are optional :
+
+| field | description     |
+|-------|--------------------------------------|
+|simPin          |If your sim card is protected, fill this field with the PIN code|
+|pollingInterval| Delay between two checks for new message|
+|delayBetweenSend|Delay to wait between two messages post (could be necessary for slow modem)|
```
You could specify the type of the parameter (String, Decimal, ...) and the unit (seconds, ...).

dalgwen replied:
Indeed ! Done

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
````diff
@@ -0,0 +1,129 @@
+|pollingInterval| Delay between two checks for new message|
+|delayBetweenSend|Delay to wait between two messages post (could be necessary for slow modem)|
+
+```
+Bridge smsmodem:smsmodembridge:adonglename [ serialPortOrIP="/dev/ttyUSB0", baudOrNetworkPort="19200", enableAutoDiscovery="true" ]
+```
````
This should go in the Full Example section.

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
```diff
@@ -0,0 +1,129 @@
+
+The *smsconversation* thing is just a shortcut to address/receive messages with a specific msisdn. It is not mandatory to use the binding, as you can use action and trigger channel to send/receive a message once the smsmodem bridge is configured.
+
+| field | description               |
```
```suggestion
| Parameter Name | description               |
```

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
```diff
@@ -0,0 +1,129 @@
+
+| field | description               |
+|-------|--------------------------|
+| recipient | The msisdn of the phone you want to discuss with.|
+| deliveryReport | If enabled, ask the network for a delivery report (default false)|
+| encoding | The encoding to use when sending the message (either Enc7, Enc8, EncUcs2, EncCustom, default is Enc7). EncUcs2 is good for non latin character, but SMS character size limit is then reduced|
```
Please specify the types of the parameters.

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
```diff
@@ -0,0 +1,129 @@
+The *smsmodembridge* has the following trigger channel :
+| Channel ID          | event                      |  
+|---------------------|----------------------------|
+|receivetrigger| The msisdn and message received (concatened with the '\|' character as a separator)|
```
"trigger" is redundant and can therefore be removed.
```suggestion
|receive| The msisdn and message received (concatened with the '\|' character as a separator)|
```

dalgwen replied:
There is already a "receive" channel. So you are saying, there won't be a naming issue ?

fwolter replied:
Ah, seems like I overlooked this. Then, everything is fine.

fwolter commented on bundles/org.openhab.binding.smsmodem/README.md:
```diff
@@ -0,0 +1,129 @@
+
+### Send SMS
+
+`sms.rules` for DSL :
```
```suggestion
`sms.rules` for DSL:
```

dalgwen replied:
Done. Why french and english can't have the same rules ??? 😅

fwolter replied:
Interesting. Wasn't aware of that fine difference.

fwolter reviewed (changes requested):
I reviewed the readme so far.

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html` after building it with `mvn clean install`.

dalgwen commented:
Thank you @fwolter !
I checked the code-analysis but the report says there is no issue now.
Do you still have some ?

fwolter commented:
> I checked the code-analysis but the report says there is no issue now.

Looks good! Just give me a note if you want to change the modelling of the bridge or whether I shall continue reviewing.

dalgwen commented:
Thank you @fwolter 
You know better than me the good usage 👍, I will change the modelling of the bridge

dalgwen commented:
As proposed, I split the bridge thing-type in two different bridge thing-types.
Binding is now ready for review.

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSConversationHandler.java:
```diff
@@ -0,0 +1,152 @@
+        setStatusByBridgeStatus();
+    }
+
+    private void setStatusByBridgeStatus() {
+        SMSModemBridgeHandler bridgeHandlerFinal = bridgeHandler;
+        if (bridgeHandlerFinal != null) {
+            switch (bridgeHandlerFinal.getThing().getStatus()) {
+                case INITIALIZING:
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+                    break;
+                case OFFLINE:
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+                    break;
+                case ONLINE:
+                    updateStatus(ThingStatus.ONLINE, ThingStatusDetail.NONE);
+                    break;
+                case REMOVED:
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+                    break;
+                case REMOVING:
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+                    break;
+                case UNINITIALIZED:
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+                    break;
+                case UNKNOWN:
+                    updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.BRIDGE_OFFLINE);
+                    break;
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+        }
```
The Thing status is set by the framework automatically according the bridge status, if you set the Thing status to UNKNOWN in `initialize()`. Then, this could be removed.

dalgwen replied:
I made a test and if I set the child handler status to "UNKNOWN" and leave if for the Bridge to update (deleting setStatusByBridgeStatus() method), it stays "UNKNOWN", even if the bridge status switch to "online". 
Or maybe I didn't understand your proposition ?

By reading code, the bridgeStatusChanged() method of BaseThingHandler effectively change the child handler status to "ONLINE" when the bridge get online, but only if the last status details was BRIDGE_OFFLINE

fwolter replied:
You still need to set the child Thing ONLINE, if you think it is online. In your case this might be always the case, if the bridge is online. So, you can set it ONLINE in `initialize()` directly. Then, it will be online if the bridge is online and offline if the bridge is offline.

You only need to set it to UNKNOWN (and later to ONLINE) if you don't know the state in `initialize()`, yet.

dalgwen replied:
Ah, I get it now !
Indeed, I can set it to ONLINE directly.
Done, thanks

By the way, there is one drawback : the bridge start as "unknown" because it has to establish connection. But the smsconversation thing is immediately online at this time (and the status doesn't reflect the reality : it is not fully functionnal because of the bridge).

fwolter replied:
Just tested your code and a reason why it doesn't work as expected is that you don't set the ThingStatus in the bridge's `initialize()` to UNKNOWN. It works for me if I do that.

dalgwen replied:
![image](https://user-images.githubusercontent.com/1935721/203526641-ab20979e-5b37-4803-99e0-9ecf665fe7e4.png)

Strange, it doesn't work for me. When the bridge is initializing (unknown), the child conversation thing is online. Maybe an order initialization issue ?
But it is no big deal, I'm OK with that, so I let you decide if it needs further investigation.

fwolter replied:
I just tested it again and you're right, as long as the bridge is in UNKNOWN state, the child Things are ONLINE. Only if the bridge is set to OFFLINE, the child Things are set to OFFLINE, too. Normally you wouldn't notice, because the bridge stays in UNKNOWN state only for very short time. In this case the timeout is quite long.

I would expect the SMS modem to respond within subseconds and would chose a shorter timeout, anyway. But maybe there are things I didn't consider.

I merge it for now. If you want to improve the behavior, you could file a follow-up PR.

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,473 @@
+    /**
+     * The smslib object responsible for the serial communication with the modem
+     */
+    private @NonNullByDefault({}) Modem modem;
+
+    /**
+     * A scheduled watchdog check
+     */
+    private @NonNullByDefault({}) ScheduledFuture<?> checkScheduled;
```
The null check annotations should only be disabled in certain cases like injection. These `@NonNullByDefault({})` should be removed and the attributes be checked for null values before they are accessed.

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,473 @@
+        modem = null;
+    }
+
+    public SMSModemBridgeHandler(Bridge bridge, SerialPortManager serialPortManager) {
```
The constructur should be placed after the field declerations.

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,473 @@
+            }
+        } catch (ModemConfigurationException e) {
+            String message = e.getMessage();
+            logger.error(message, e);
```
The log message could be removed, as the status update is already logged by the framework.

The state change originated by updateStatus() is logged to events.log. Including the status detail message.

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,473 @@
+        } catch (ModemConfigurationException e) {
+            String message = e.getMessage();
+            logger.error(message, e);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, message);
```
Is this rather a CONFIGURATION_ERROR?

dalgwen replied:
Both !
A misconfiguration (bad address) or network/serial error prevents from communicating with the modem, which is not joinable at the address set.
So, what should I use ?

fwolter replied:
This is what I saw. That sounds more like a configuration error. I leave it up to you what makes more sense.
![grafik](https://user-images.githubusercontent.com/1894841/202898901-0b5f44ea-0a30-4517-bac4-364ae0c7b677.png)

dalgwen replied:
You are right.
I splitted it in two, depending on the root cause (IOException = communication error, other = configuration error).
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,473 @@
+            if (text != null) {
+                messageText = text;
+            } else {
+                logger.error("Message has no payload !");
```
See above. Logging to error. Please check all.

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,473 @@
+        return modem != null && (modem.getStatus() == Status.Started || modem.getStatus() == Status.Starting);
+    }
+
+    @Override
+    public void childHandlerInitialized(ThingHandler childHandler, Thing childThing) {
+        if (childHandler instanceof SMSConversationHandler) {
+            childHandlers.add((SMSConversationHandler) childHandler);
+        } else {
+            logger.error("The SMSModemBridgeHandler can only handle SMSConversation as child");
+        }
+    }
```
Are you aware you can access the child Things via `getBridge().getThings()`?

dalgwen replied:
Thank you, modification done.

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,473 @@
+    public boolean processStatusCallback(Modem.Status oldStatus, Modem.Status newStatus) {
+        switch (newStatus) {
+            case Error:
+                logger.error("SMSLib reported an error on the underlying modem {}", modem.getDescription());
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
You could add the error message as a third argument to `updateStatus()`. Then, the error message is displayed in the UI and the logging could be removed.

dalgwen replied:
Thanks, done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/resources/OH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>SMSModem Binding</name>
+	<description>This binding handle a GSM modem connected to the openHAB server (Serial), or exposed on the network. It
```
```suggestion
	<description>This binding handles a GSM modem connected to the openHAB server (Serial), or exposed on the network. It
```

dalgwen replied:
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/resources/OH-INF/thing/smsmodem.xml:
```diff
@@ -0,0 +1,91 @@
+		</config-description>
+	</bridge-type>
+
+	<channel-type id="smsmodemreceivetrigger">
+		<kind>trigger</kind>
+		<label>Message Received</label>
+		<description>Triggered when a message is received, in the form "&lt;msisdn_sender&gt;|&lt;text&gt;"</description>
+		<event>
+		</event>
+	</channel-type>
```
Can you apply the correct indentation?

dalgwen replied:
Sorry, I don't see what is the issue here. Is it for the "event" element ? if so, I changed it to <event/>

dalgwen replied:
My answer was truncated.
Sorry, I don't see what is the issue here. Is it for the "event" element ? if so, I changed it to <event/>

dalgwen commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/SMSModemHandlerFactory.java:
```diff
@@ -0,0 +1,105 @@
+    }
+
+    public static ThingUID getSMSConversationUID(ThingTypeUID thingTypeUID, String recipient, ThingUID bridgeUID) {
+        return new ThingUID(thingTypeUID, recipient, bridgeUID.getId());
```
The "getSMSConversationUID" method is called by the autodiscovery method, or from the createThing method if and only if the "thingUID" is null.
With this method, I want to give a default UID, IF NOT set by the user. If the user set the thing UID, it will use the user's one.
Your comment seems to mean that this approach is invalid, or do you mean it's badly implemented ?

fwolter replied:
In `SMSConversationDiscoveryService` you are calling `getSMSConversationUID()` directly, when creating the discovered Thing (which is the right approach). So, the default-ThingUID-use-case is handled there.

If the user adds the Thing manually, the ThingUID can't be left empty/null (neither via UI nor text based configuration).

What use case do you intend to handle by this code?

dalgwen replied:
No particular use case here. I was just trying to implement the createThing method from the parent class and respecting the annotation : in this method, the parameter thingUID is nullable.
If you thing it is best, I can remove the call to the getSMSConversationUID in createThing, but the compiler will log a warning I guess ?
Or I could fail fast with an exception if thingUID is null.
What do you think ?

dalgwen replied:
Oh my ! I wouldn't have wasted your time if I knew that before... Sorry 😨
Deleted !

dalgwen commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/SMSModemBridgeConfiguration.java:
```diff
@@ -0,0 +1,30 @@
+public class SMSModemBridgeConfiguration {
+
+    public String serialPort = "";
+    public Integer baud = 9800;
```
Yes, indeed, 9600 seems more valid than 9800.
(Mine is 19200, I don't know what is the most used value.)

Done

dalgwen commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/actions/SMSModemActions.java:
```diff
@@ -0,0 +1,81 @@
+        if (recipient != null && !recipient.isEmpty() && message != null) {
+            handler.send(recipient, message, false, encoding);
+        } else {
+            logger.error("SMSModem cannot send a message with no recipient or text");
```
Thanks. (and sorry, I keep making the same mistake despite being told each time 😂... )

Done

dalgwen commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSConversationHandler.java:
```diff
@@ -0,0 +1,152 @@
+            Bridge bridge = getBridge();
+            if (bridge == null) {
+                logger.error("Required bridge not defined for SMSconversation {} with {}.", thing.getUID(),
+                        getRecipient());
```
Done

fwolter commented on bundles/org.openhab.binding.smsmodem/src/main/java/org/openhab/binding/smsmodem/internal/handler/SMSModemBridgeHandler.java:
```diff
@@ -0,0 +1,487 @@
+
+    @Override
+    public void initialize() {
+        shouldRun = true;
```
This is necessary to make the thing status automatism work.
```suggestion
        updateStatus(ThingStatus.UNKNOWN);
        shouldRun = true;
```

fwolter reviewed (approved):
LGTM
Now, you could add your binding's logo to the openHAB website. See https://www.openhab.org/docs/developer/bindings/#add-your-binding-s-logo-to-the-openhab-website

dalgwen commented:
Great, thank you again for your time !

lolodomo commented:
As the French translations were included on the PR, they have to be uploaded manually in Crowdin.

lolodomo commented:
@dalgwen : you forgot adding an entry in CODEOWNERS file, can you please submit a small PR for that?

dalgwen commented:
> As the French translations were included on the PR, they have to be uploaded manually in Crowdin.

First time in crowdin for me. I think I did what you suggest. With some fix after viewing warning from crowdin.

> you forgot adding an entry in CODEOWNERS file, can you please submit a small PR for that?

Oops, sorry. (Done here)[https://github.com/openhab/openhab-addons/pull/13834].

lolodomo commented:
> First time in crowdin for me. I think I did what you suggest. With some fixes after viewing warnings from crowdin.

I have validated with few changes.
Syncing is in progress, the PR should be created in few minutes.
